common/ssl_util.c

```diff
@@ -1090,6 +1089,210 @@ ssl_verify_signature(const char *cert_file, const char *signature_file, const ch
+		goto error;
+	}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+	EVP_MD_CTX _md_ctx;
+	md_ctx = &_md_ctx;
+	EVP_MD_CTX_init(md_ctx);
+#else
+	if ((md_ctx = EVP_MD_CTX_new()) == NULL) {
+		ERROR("Allocating EVP_MD failed!");
+		goto error;
+	}
+#endif
+	EVP_VerifyInit(md_ctx, hash_fct);
+
+	if (!EVP_VerifyUpdate(md_ctx, buf, buf_len)) {
+		ERROR("Error in signature verification (reading/hashing signed file failed");
+		ret = -2;
+		goto error;
+	}
+
+	DEBUG("File hash computed to verify signature");
+
+	ret = EVP_VerifyFinal(md_ctx, sig_buf, sig_len, key);
+	if (ret != 1) {
+		DEBUG("Signature verification error");
+		// any error
+		if (ret == -1)
+			ret = -2;
+		// verification failed
+		else
+			ret = -1;
+	} else {
+		DEBUG("Signature successfully verified");
+		ret = 0;
+	}
+
+error:
+	if (cert)
+		X509_free(cert);
+	if (key)
+		EVP_PKEY_free(key);
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+	EVP_MD_CTX_cleanup(md_ctx);
+#else
+	if (md_ctx)
+		EVP_MD_CTX_free(md_ctx);
+#endif
+	return ret;
+}
+
+int
+ssl_verify_signature_from_digest(const char *cert_buf, const uint8_t *sig_buf, size_t sig_len,
+				 const uint8_t *hash, size_t hash_len)
+{
+	ASSERT(cert_buf);
+	ASSERT(sig_buf);
+	ASSERT(hash);
+
+	int ret = 0;
+	X509 *cert;
+	EVP_PKEY *key = NULL;
+	BIO *mem;
+	EVP_PKEY_CTX *pkey_ctx = NULL;
+	const EVP_MD *hash_fct;
+
+	// load certificate, verify and get public key
+	mem = BIO_new(BIO_s_mem());
+	BIO_puts(mem, cert_buf);
+	cert = PEM_read_bio_X509(mem, NULL, 0, NULL);
+	BIO_free(mem);
+
+	if ((key = X509_get_pubkey(cert)) == NULL) {
+		ERROR("Error in signature verification (loading pubkey failed)");
+		ret = -2;
+		goto error;
+	}
+
+	const X509_ALGOR *sig_alg = X509_get0_tbs_sigalg((const X509 *)cert);
+	if (!sig_alg) {
+		ERROR("Error in signature verification (Failed to parse hash-algorithm)");
+		ret = -2;
+		goto error;
+	}
+
+	const char *hash_algo = asn1_object_to_hash_algo(sig_alg->algorithm);
+	if (!hash_algo) {
+		ERROR("Error in signature verification (Unsupported hash function)");
+		ret = -2;
+		goto error;
+	}
+
+	if ((hash_fct = EVP_get_digestbyname(hash_algo)) == NULL) {
+		ERROR("Error in signature verification (unable to initialize hash function)");
+		ret = -2;
+		goto error;
+	}
+
+	DEBUG("Verifying signature...");
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+	EVP_PKEY_CTX _pkey_ctx;
+	pkey_ctx = &_pkey_ctx;
+	EVP_PKEY_CTX_init(md_ctx);
+#else
+	if ((pkey_ctx = EVP_PKEY_CTX_new(key, NULL)) == NULL) {
+		ERROR("Allocating EVP_PKEY_CTX failed!");
+		ret = -2;
+		goto error;
+	}
+#endif
+
+	ret = EVP_PKEY_verify_init(pkey_ctx);
+	if (ret != 1) {
+		ret = -2;
+		ERROR("EVP_PKEY_verify_init failed");
+		goto error;
+	}
+
+	if (EVP_PKEY_CTX_set_signature_md(pkey_ctx, EVP_sha256()) != 1) {
+		DEBUG("EVP_PKEY_CTX_set_signature_md failed");
+		ret = -2;
+		goto error;
+	}
+
+	ret = EVP_PKEY_verify(pkey_ctx, sig_buf, sig_len, hash, hash_len);
+	if (ret != 1) {
+		DEBUG("EVP_PKEY_verify error");
+		// any error
+		if (ret == -1) {
```

Comment:
same here, this does not make sense since in if block (ret != 1)

Reply:
Done as discussed

Reply:
i missed the '-' sorry for the noise.